### ISSUE TITLE
Have dnf_package support modular/appstream installs

### ISSUE THREAD

**Issue by erinn** [chef/chef#9220]
### Describe the Enhancement:
I'd like to see the dnf_package resource support modules/appstreams for RHEL 8+ and fedora

### Describe the Need:
Well modules/appstreams solve a large need for users who want to use RHEL/fedora but want most things stable and a few things changed to a newer version. A good example here is PHP, if you just use the built in version of PHP in RHEL you are locked into the stone age pretty quickly. modules/appstreams solve this by allowing newer versions (supported by RHEL if you pay them) of PHP to be installed.

### Current Alternative
execute I supose: execute 'yum -y module install php:7.3/devel' etc. 

### Can We Help You Implement This?:
Well yes you can.

**Comment by lamont-granquist** [comment]
Have you tried `dnf_package "php:7.3/devel"` and what does that do?  (and what relevant logging comes out if you use -l trace)?

**Comment by erinn** [comment]
Well with just dnf_package "php:7.3/devel" you get this: No candidate version available for php:7.3/devel

Abusing options like so:
dnf_package "php:7.3/devel" do
  options 'module'
end
Creates the command:
`dnf module -y install php:7.3/devel`
Which looks a little funny but will actually work from the CLI, chef outputs the same result. My suspicion, uninvestigated at this point, is that the python helper is catching this issue. But this is based off of my knowledge of yum_package, dnf_package may be entirely different. 

I also have to warn that this is being tested on chef infra 14, because of the licensing changes I can't test on chef infra 15.

**Comment by lamont-granquist** [comment]
ah i see i missed the module bit of `yum module install`

**Comment by lamont-granquist** [comment]
and yeah this would take surgery to the python helper to know anything about modules

**Comment by lamont-granquist** [comment]
this should probably be considered blocked on #9013 since any work done to the dnf_package provider to support additional features will only make doing that work exponentially harder.

**Comment by ramereth** [comment]
I'd like to revisit this now that #9013 was closed.

I think it might be useful to also create a new resource called ``dnf_module`` as users sometimes need to enable/disable modules such as what's described [here](https://www.percona.com/doc/percona-server/LATEST/installation/yum_repo.html).

Looking at the help for ``dnf module --help``, you can see the following actions:
```
  <modular command>     disable: disable a module with all its streams
                        enable: enable a module stream
                        info: print detailed information about a module
                        install: install a module profile including its packages
                        list: list all module streams, profiles and states
                        provides: list modular packages
                        remove: remove installed module profiles and their packages
                        repoquery: list packages belonging to a module
                        reset: reset a module
                        update: update packages associated with an active stream
```
The backend actually creates/manages files under ``/etc/dnf/modules.d`` similar to yum repos. I'm not sure if this resource should directly manage those files or do it via ``dnf module`` commands.

I'm also not sure how this would tie in directly into the ``dnf_package`` resource.

**Comment by ramereth** [comment]
I think for this specific use case you'd want to do something like this:
```ruby
dnf_module 'php:7.3' do
  action :enable
end

package 'php-devel'
```
Perhaps it might make sense to have ``:enable`` be the default action. Here's what it looks like from the CLI if you do it manually:
```
$ dnf module enable php:7.3
Last metadata expiration check: 0:00:32 ago on Mon 03 Aug 2020 05:36:22 PM UTC.
Dependencies resolved.
=======================================================================================================================
 Package                     Architecture               Version                      Repository                   Size
=======================================================================================================================
Enabling module streams:
 httpd                                                  2.4
 nginx                                                  1.14                                                          
 php                                                    7.3                                                           

Transaction Summary
=======================================================================================================================

Is this ok [y/N]: y
Complete!
```

**Comment by lamont-granquist** [comment]
Yeah `dnf_package` is about installing a package or an array of single packages.  Installing modules should go into a different resource.  Single Responsibility Principle and the dnf_package provider is already overly complicated for what it does and does not need more functionality surgically bolted onto it.

**Comment by Annih** [comment]
Hello @ramereth , the new `dnf_module` resource is quite good, do you know if there is any plan to make it a native resource?

**Comment by thorhs** [comment]
I would also be interested in a dnf_module resource be in the chef proper.
